examples/tutorial_4_streamproto2.md

```diff
@@ -0,0 +1,106 @@
+In the [previous tutorial](tutorial_3_streamproto.md), we learned how to create a custom streaming protocol.
+
+This tutorial is very similar to the [previous tutorial](tutorial_3_streamproto.md) in that a stream of data will be consumed, however in this tutorial we will learn how to use alternative libp2p streaming semantics.
+
+# Custom protocol in libp2p
+Let's create a `part4.nim`, and import our dependencies:
+```nim
+import bearssl
+import chronos
+import stew/byteutils
+
+import libp2p
+```
+This is similar to the third tutorial. Next, we'll declare our custom protocol.
+```nim
+const TestStreamCodec = "/test/stream-proto/1.0.0"
+
+type TestStreamProto = ref object of LPProtocol
+```
+
+Just as we did in the last tutorial, we've set a [protocol ID](https://docs.libp2p.io/concepts/protocols/#protocol-ids), and created a custom `LPProtocol`.
+
+As in the last tutorial, we are going to handle our client and server parts separately. The server will wait for an instruction from the client, then send a stream of data to the client and finish by closing the connection. The client will listen in a loop until the server has finished and closed the connection. Then the client will close its end of the connection as well.
+
+Let's start with the server part:
+```nim
+# Set the chunk length >= any data being streamed. Our longest chunk of data in
+# either direction is 16 bytes ("please send data").
+const CHUNK_LEN_BYTES = 16
+
+proc new(T: typedesc[TestStreamProto]): T =
+  # every incoming connections will in be handled in this closure
+  var buf = newSeq[byte](CHUNK_LEN_BYTES)
+  proc handle(stream: Connection, proto: string) {.async, gcsafe.} =
+
+    discard await stream.readOnce(addr buf[0], CHUNK_LEN_BYTES)
+    echo "[server] received from client: ", string.fromBytes(buf)
+
+    # send stream of data
+    for i in 0..10:
+      let data = "Part " & $i
+      await stream.write(data)
+
+    # We must close the connections ourselves when we're done with it
+    await stream.close()
+
+  return T(codecs: @[TestStreamCodec], handler: handle)
+```
+Instead of using `readLp` to read data from the connection (as in the previous tutorial), we are going to use `readOnce`, which allows us to read bytes from the buffer or wait for bytes to be received.
+
+We can now create our client part:
+```nim
+proc streamData(p: TestStreamProto, stream: Connection) {.async.} =
+  let data = "please send data"
+  await stream.write(data)
+
+  # Read loop
+  while stream.atEof.not:
+
+    var buf = newSeq[byte](CHUNK_LEN_BYTES)
+    # readOnce reads bytes from internal buffer if present, otherwise waits for
+    # for bytes to be received.
```

Comment:
```suggestion
    # for bytes to be received. Note that it will return as soon as it receive bytes, even if the message is
    # not complete.
```

Or maybe I would use readExactly here, actually?

Reply:
`readExactly` will throw if not all bytes have been consumed or hang if not enough have arrived. You usually want to use `readExactly` only when you know exactly how many bytes are coming, for example with fixed chunk sizes or length prefixed messages (a la `readLp/writeLp`)

Reply:
In tutorial 4, we aren't using `readLp/writeLp` though. And sometimes we send 16 bytes (`please send data`), sometimes 6 bytes (eg `Part 1` ... `Part 9`), and sometimes 7 bytes (eg `Part 10`). So the length isn't exactly known, right? Changing `readOnce` to `readExactly` gets this in the output:
```
[server] received from client: please send data
[client] received from server: Part 0Part 1Part
[client] received from server:  2Part 3Part 4Pa
[client] received from server: rt 5Part 6Part 7
[client] received from server: Part 8Part 9Part
```
followed by an `LPStreamIncompleteError` exception.

Reply:
That was a cool experiment to see what the difference is between `readOnce` and `readExactly`. Would make a nice tutorial 😂

Reply:
Yes, to use readExactly you have to send fixed size messages :)

Could indeed be a nice thing to have a tutorial showing the 3 methods (readOnce, readExactly and readLp) and their usecases/drawbacks (maybe one tutorial for the 3 would be enough, bombarding with  3tutorials for the same concept might be a bit too much)

But we can do that later

However it should be noted that with readOnce, you don't know if you'll get entire messages

eg, if you sleep in the receiver, it will probably receive every message at once, I think?

Reply:
sleeping doesn't guarantee that you get all data - with readOnce, you roughly get one "buffer" at a time, but there are many buffers of many sizes along the way so you can't tell - when you sleep, those buffers can become full so you'll never get "everything" guaranteed

Reply:
Uh, indeed. Interestingly, in this scenario, since "lower layers" (notably the multiplexer) are packet based, if you call readOnce, you only get one packet, even if there is more bytes queued up.

I hope that nobody builds on top of this assumption, because that's a very shaky and hard to debug one.

I would even be inclined to remove it, by doing this:
```diff
--- a/libp2p/stream/bufferstream.nim
+++ b/libp2p/stream/bufferstream.nim
@@ -128,7 +128,7 @@ method readOnce*(s: BufferStream,
   # First consume leftovers from previous read
   var rbytes = s.readBuf.consumeTo(toOpenArray(p, 0, nbytes - 1))
 
-  if rbytes < nbytes and not s.isEof:
+  while rbytes < nbytes and not s.isEof:
     # There's space in the buffer - consume some data from the read queue
     s.reading = true
     let buf =
@@ -158,6 +158,8 @@ method readOnce*(s: BufferStream,
         trace "add leftovers", s, len = buf.len - remaining
         s.readBuf.add(buf.toOpenArray(remaining, buf.high))
 
+    if s.readQueue.len() == 0: break
+
   if s.isEof and s.readBuf.len() == 0:
     # We can clear the readBuf memory since it won't be used any more
     s.readBuf = StreamSeq()
```

Which gives for the tutorial, on my machine, in this day and age:
```
[server] received from client: please send data
[client] received from server: Part 0Part 1
[client] received from server: Part 2
[client] received from server: Part 3Part 4
[client] received from server: Part 5
[client] received from server: Part 6Part 7
[client] received from server: Part 8
[client] received from server: Part 9Part 10
[client] received from server: 
```
(It's still kinda packet delimited, because the readQueue in the BufferStream is packet based instead of just a stream of bytes)

Thankfully I think almost everyone uses the length prefixed variants, so it shouldn't break anything, but since this kind of bugs are a real plague: very timing dependent and random

And while we can "guarantee" that with mplex, readOnce will give you a packet, it will not be the case with other multiplexers

Reply:
> eg, if you sleep in the receiver, it will probably receive every message at once, I think?

Not really, as noted by @arnetheduck you get roughly one buffer in flight at any given time and with back pressure, you shouldn't end up in a situation where buffers overflow. It is true that this buffers are usually message based, because this is how mplex (and most other libp2p protocols) work, but you still get backpressure at the message boundaries. In other words (and you should be able to verify this), the writer won't be able to send any more data unless the read has picked up the last message from the stream.

I'm not sure what issues you're running in the above reply, but the changes don't seem correct, can you describe the behavior a little better?